dialer.go

```diff
@@ -0,0 +1,24 @@
+package mesg
+
+import (
+	"net"
+	"time"
+)
+
+// Dialer describes a dialer implementation for network.
+type Dialer interface {
+	Dial() (net.Conn, error)
+}
+
+type gRPCDialer struct {
+	dialer Dialer
+}
+
+func newGRPCDialer(dialer Dialer) *gRPCDialer {
+	return &gRPCDialer{dialer}
+}
+
+// Dial used to produce a net.Conn to connect gRPC server with.
+func (d *gRPCDialer) Dial(addr string, timeout time.Duration) (net.Conn, error) {
+	return d.dialer.Dial()
```

Comment:
shouldn't we pass the arguments `addr` and `timeout` here ?  I'm not sure I really understand why this method is here

Reply:
mesgtest package returns an in memory connection [here](https://github.com/ilgooz/go-application/blob/feature/init/mesgtest/server.go#L36) and returned _*Socket_ implements [Dialer](https://github.com/ilgooz/go-application/blob/5bf6c9e/dialer.go#L9). We get that Dialer with application Option returned from [DialOption](https://github.com/ilgooz/go-application/blob/2fd575e/application.go#L77). This way we're able to provide a fake in-memory connection between application and fake server.

While I'm defining Dialer interface I was only thinking about to use it with mesgtest, that's why I haven't added _addr_ and _timeout_ arguments to our custom Dialer. Because we don't need them while testing, since it uses an in-memory net connection.

But, gRPC's [WithDialer()](https://github.com/ilgooz/go-application/blob/feature/init/application.go#L79) expects a Dial function with addr and timeout arguments, like the method you put your comment on. To make gRPC's Dial function to  be compatible with our Dialer, I've created _gRPCDialer_ type and put a gRPC Dial compatible method to it that calls our Dialer's _Dial()_ method.

Also, I don't think that our Dialer should have a Dial method that accepts an _addr_ and _timeout_ to let devs to provide a custom type that implements Dialer. Because they don't need to create a tcp connection with mesg-core manually. It would only make sense to provide this kinda flexibility if mesg-core weren't running as a standalone  program and just be a dependency package. Because that way devs can create a multiplexed or some kinda special connection with their server and provide a custom connection to both go-application in the client side and mesg-core in the server side...

As a note, gRPC's Dial function should be expecting _addr_ and _timeout_ to provide values to custom Dial function with values it got from `grpc.DialContext()`. In our case, we don't need to use them.

Reply:
thanks for the explanation. Got it 👍